keps/sig-scheduling/3838-pod-mutable-scheduling-directives/README.md

```diff
@@ -0,0 +1,950 @@
+
+The main use case for this feature was to enable building external resource controllers (like extended schedulers, 
+dynamic quota managers) to make a decision on when the pod should be eligible for scheduling by kube-scheduler.
+
+This enhancement proposes to make pod scheduling directives (nodeSelector, affinity) mutable as long as the pod
+is gated and the update only further constrains the scheduling of the pod, and doesn't expand it.
+
+This allows external resource controllers to not only decide when the pod should be eligible for scheduler, but 
+also where it should land. This is similar to what we did for suspended Job in [#2926](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/2926-job-mutable-scheduling-directives); 
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Adding the ability to mutate a Pod's scheduling directives before it is allowed to schedule 
+gives an external resource controller (like a network toplogy-aware scheduler or a dynamic 
+quota managers) the ability to influence placement while at the same time offloading actual 
+pod-to-node assignment to the existing kube-scheduler.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Allow adding node affinity or node selector of a pod that is blocked on a
+  scheduling readiness gate.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Allow mutating scheduling directives of pods unconditionally.
+- Restrict updating tolerations on scheduling gates. We already allow adding tolerations 
+  unconditionally, and so restricting that to gated pods is a backward incompatible change.
```

Comment:
Did we allow "add" or "update" for tolerations changes of created pods? (better make the wordings consistent)

Reply:
We allow adding only.

Reply:
done.